### PR TITLE
go: add log messages guide

### PR DESCRIPTION
**Pull request by kopiczko**

### CODE REVIEW

File: go/log_messages.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+```go
+var err error
+
+// Note all messages contain "system fact" which is human readable name for the
+// system fact being discovered. Exception is "cancelling resource" message.
+var systemFact string
+{
+	// Start message. Note "finding out" here.
+	r.logger.LogCtx(ctx, "level", "debug", "message", "finding out system fact")
```
Comment (kopiczko) [inline go/log_messages.md]: I thought about good verb here. Like `get`, `look for`, `figure`. I think `find` works well with our `IsNotFound` matchers.

File: go/log_messages.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+	systemFact, err = pkg.FindSystemFact(ctx)
+	if IsNotFound(err) {
+		// Not found log. Note "did not find" here.
+		r.logger.LogCtx(ctx, "level", "debug", "message", "did not find system fact")
+		// Reason log should be separate free form log explaining why
+                // the system fact can't be found.
+		r.logger.LogCtx(ctx, "level", "debug", "message", "waiting for system fact to be created")
```
Comment (kopiczko) [inline go/log_messages.md]: Added a reason block which serves as an explanation for for the dry log of not being able to find info about the fact.

File: go/log_messages.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+		// Canceling resource log should be separate.
+		r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
```
Comment (kopiczko) [inline go/log_messages.md]: This is deviation of what we have now in many places but `canceling resource reconciliation for custom object` contains redundant or incorrect information. E.g. when the controller is watching pods we don't reconcile custom objects.

**Review by kopiczko — COMMENTED** [review]
This is just something I made to achieve the consensus. Not everything here is created as I like it to be, but I tried to find common ground based on what I saw in the reviews and discussion so far. Let's talk!

File: go/log_messages.md (hunk 3 above)
Comment (xh3b4sd) [reply go/log_messages.md]: Makes sense. When the whole reconciliation is canceled it should then simply be `canceling reconciliation`.

File: go/log_messages.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+{
+	// Start message. Note "ensured" here. "deletion of" should be present
+	// in case of delete event.
+	r.logger.LogCtx(ctx, "level", "debug", "message", "ensuring [deletion of] managed resource")
+
+	err = pkg.APICall(ctx, systemFact)
+	if IsDeletionInProgress(err) {
+		// Did not ensure log. Note "did not ensure" here. "deletion"
+		// should be present in case of delete event.
+		r.logger.LogCtx(ctx, "level", "debug", "message", "did not ensure [deletion] managed resource")
```
Comment (xh3b4sd) [inline go/log_messages.md]: `[deletion]` -> `[deletion of]`

**Review by xh3b4sd — APPROVED** [review]
This is really cool. I like it a lot.

File: go/log_messages.md (hunk 2 above)
Comment (indyfree) [reply go/log_messages.md]: Nitpick: There is an extra space in line 34.

File: go/log_messages.md (hunk 3 above)
Comment (indyfree) [reply go/log_messages.md]: For me `canceling resource reconciliation` is the most intuitive. Without the `reconsiliation` it sounds a bit odd to me, but no strong opinions.

**Review by indyfree — APPROVED** [review]
Super cool. Makes total sense to align this!

File: go/log_messages.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
+## Log Messages
+
+### Resource Reconciliation
+
+When doing resource reconciliation action we often have to find out the current
+state, and apply changes to bring it towards desired state. This often means
+calling some API and then emitting debug logs about that. We should try to
```
Comment (MarcelMue) [inline go/log_messages.md]: I think this is the most important part and not limited to reconciliation. We call APIs in other places as well and surrounding our calls with this kind of logging is not only helpful for operators IMO.

**Review by MarcelMue — APPROVED** [review]
I like it, we can iterate on this later.

File: go/log_messages.md (hunk 3 above)
Comment (xh3b4sd) [reply go/log_messages.md]: @indyfree how do you express the cancelation of the whole loop then? We can either cancel only one resource within the loop or the whole loop.

**Review by rossf7 — APPROVED** [review]
LGTM. Will make reading our logs easier which is really cool. <3

File: go/log_messages.md (hunk 3 above)
Comment (kopiczko) [reply go/log_messages.md]: Looks like `canceling resource` and `canceling reconciliation` wins here.
We should also write down messages for other context related stuff, like keeping finalizers. But let's make it separate PR.

**Comment by kopiczko** [comment]
Last call. Merging tomorrow morning.
I think we should update with reconciliation canceling keeping finalizers, and other potential context packages logs. This will be in next PR to not make too many discussions at once.

**Pull request merged**